File: src/main/java/main/commands/prefixcommands/audio/LoopCommand.java
package main.commands.prefixcommands.audio;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import lavalink.client.player.IPlayer;
import main.audiohandlers.RobertifyAudioManager;
import main.audiohandlers.GuildMusicManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.json.logs.LogType;
import main.utils.json.logs.LogUtils;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.annotations.ForRemoval;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.internal.utils.tuple.Pair;

import javax.script.ScriptException;
import java.util.List;

@Deprecated @ForRemoval
public class LoopCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final Message msg = ctx.getMessage();
        final Member self = ctx.getSelfMember();
        GuildVoiceState selfVoiceState = self.getVoiceState();

        EmbedBuilder eb;

        final Member member = ctx.getMember();
        final GuildVoiceState memberVoiceState = member.getVoiceState();

        var musicManager = RobertifyAudioManager.getInstance().getMusicManager(ctx.getGuild());
        var audioPlayer = musicManager.getPlayer();

        if (checks(selfVoiceState, memberVoiceState, audioPlayer) != null) {
            msg.replyEmbeds(checks(selfVoiceState, memberVoiceState, audioPlayer).build()).queue();
            return;
        }

        if (ctx.getArgs().isEmpty()) {
            msg.replyEmbeds(handleRepeat(musicManager, ctx.getAuthor()).build()).queue();
            return;
        }

        if (ctx.getArgs().get(0).equalsIgnoreCase("queue") || ctx.getArgs().get(0).equalsIgnoreCase("q")) {
            eb = handleQueueRepeat(musicManager, ctx.getAuthor());
        } else {
            eb = RobertifyEmbedUtils.embedMessage(ctx.getGuild(), RobertifyLocaleMessage.GeneralMessages.INVALID_ARGS);
        }
        msg.replyEmbeds(eb.build()).queue();
    }

    public EmbedBuilder checks(GuildVoiceState selfVoiceState, GuildVoiceState memberVoiceState, IPlayer audioPlayer) {
        final var guild = selfVoiceState.getGuild();
        EmbedBuilder eb;

        if (!selfVoiceState.inAudioChannel()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);
            return eb;
        }

        if (!memberVoiceState.inAudioChannel()) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.USER_VOICE_CHANNEL_NEEDED);
            return eb;
        }

        if (!memberVoiceState.getChannel().equals(selfVoiceState.getChannel())) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()));
            return eb;
        }

        if (audioPlayer.getPlayingTrack() == null) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.LOOP_NOTHING_PLAYING);
            return eb;
        }

        if (audioPlayer.getPlayingTrack() == null) {
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.LOOP_NOTHING_PLAYING);
            return eb;
        }

        return null;
    }

    public EmbedBuilder handleRepeat(GuildMusicManager musicManager, User looper) {
        final var guild = musicManager.getGuild();

        final var player = musicManager.getPlayer();
        final var scheduler = musicManager.getScheduler();

        if (player.getPlayingTrack() == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);

        EmbedBuilder eb;

        AudioTrackInfo info = player.getPlayingTrack().getInfo();
        if (scheduler.repeating) {
            scheduler.repeating = false;
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.LOOP_STOP, Pair.of("{title}", info.title));
        } else {
            scheduler.repeating = true;
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.LOOP_STOP, Pair.of("{title}", info.title));
        }

        new LogUtils(guild).sendLog(
                LogType.TRACK_LOOP, RobertifyLocaleMessage.LoopMessages.LOOP_LOG,
                Pair.of("{user}", looper.getAsMention()),
                Pair.of("{status}", (scheduler.repeating ? "looped" : "unlooped")),
                Pair.of("{title}", info.title),
                Pair.of("{author}", info.author)

        );

        return eb;
    }

    public EmbedBuilder handleQueueRepeat(GuildMusicManager musicManager, User looper) {
        EmbedBuilder eb;
        final var scheduler = musicManager.getScheduler();
        final var guild = musicManager.getGuild();
        final var audioPlayer = musicManager.getPlayer();

        if (scheduler.playlistRepeating) {
            scheduler.playlistRepeating = false;
            scheduler.removeSavedQueue(guild);
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.QUEUE_LOOP_STOP);
        } else {
            scheduler.playlistRepeating = true;

            if (audioPlayer.getPlayingTrack() == null) {
                eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NOTHING_PLAYING);
                return eb;
            }

            AudioTrack thisTrack = audioPlayer.getPlayingTrack();

            if (scheduler.queue.isEmpty()) {
                eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.QUEUE_LOOP_NOTHING);
                return eb;
            }

            scheduler.addToBeginningOfQueue(thisTrack);
            scheduler.setSavedQueue(guild, scheduler.queue);
            scheduler.queue.remove(thisTrack);
            eb = RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.LoopMessages.QUEUE_LOOP_START);
        }

        new LogUtils(guild).sendLog(LogType.TRACK_LOOP, RobertifyLocaleMessage.LoopMessages.QUEUE_LOOP_LOG, Pair.of("{user}", looper.getAsMention()), Pair.of("{status}", (scheduler.playlistRepeating ? "looped" : "unlooped")));
        return eb;
    }

    @Override
    public String getName() {
        return "loop";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+ GeneralUtils.listToString(getAliases()) +"`\n" +
                "Set the song being currently played or the queue to constantly loop\n" +
                "\nUsage `" + prefix + "loop [queue]` *(Add `queue` to start repeating the current queue)*";
    }

    @Override
    public List<String> getAliases() {
        return List.of("l");
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/LoopCommand.kt
package main.commands.slashcommands.audio

import lavalink.client.player.IPlayer
import main.audiohandlers.GuildMusicManager
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.title
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.Command
import main.utils.component.interactions.slashcommand.models.SubCommand
import main.utils.json.logs.LogType
import main.utils.json.logs.LogUtilsKt
import main.utils.locale.messages.GeneralMessages
import main.utils.locale.messages.LoopMessages
import net.dv8tion.jda.api.entities.GuildVoiceState
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent

class LoopCommand : AbstractSlashCommand(
    Command(
        name = "loop",
        description = "Replay the current song being played.",
        subcommands = listOf(
            SubCommand(
                name = "track",
                description = "Toggle looping the song being currently played"
            ),
            SubCommand(
                name = "queue",
                description = "Toggle looping the current queue"
            )
        )
    )
) {

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!
        val musicManager = RobertifyAudioManager[guild]
        val checksEmbed = checks(event.member!!.voiceState!!, event.guild!!.selfMember.voiceState!!, musicManager.player)
        if (checksEmbed != null) {
            event.replyEmbed { checksEmbed }
                .setEphemeral(true)
                .queue()
            return
        }

        when (event.subcommandName) {
            "track" -> event.replyEmbed { handleRepeat(musicManager, event.user) }.queue()
            "queue" -> event.replyEmbed { handleQueueRepeat(musicManager, event.user) }.queue()
        }
    }

    private fun checks(
        selfVoiceState: GuildVoiceState,
        memberVoiceState: GuildVoiceState,
        player: IPlayer
    ): MessageEmbed? {
        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState)
        if (acChecks != null) return acChecks

        val guild = memberVoiceState.guild

        if (player.playingTrack == null)
            return RobertifyEmbedUtils.embedMessage(guild, LoopMessages.LOOP_NOTHING_PLAYING)
                .build()

        return null
    }

    private fun handleRepeat(musicManager: GuildMusicManager, looper: User): MessageEmbed {
        val guild = musicManager.guild
        val player = musicManager.player
        val scheduler = musicManager.scheduler
        val queueHandler = scheduler.queueHandler
        val track = player.playingTrack
            ?: return RobertifyEmbedUtils.embedMessage(
                guild,
                GeneralMessages.NOTHING_PLAYING
            ).build()

        val embed = if (queueHandler.trackRepeating) {
            queueHandler.trackRepeating = false
            RobertifyEmbedUtils.embedMessage(
                guild,
                LoopMessages.LOOP_STOP,
                Pair("{title}", track.title)
            ).build()
        } else {
            queueHandler.trackRepeating = true
            RobertifyEmbedUtils.embedMessage(
                guild,
                LoopMessages.LOOP_START,
                Pair("{title}", track.title)
            ).build()
        }

        LogUtilsKt(guild).sendLog(
            LogType.TRACK_LOOP,
            LoopMessages.LOOP_STOP,
            Pair("{user}", looper.asMention),
            Pair("{status}", if (queueHandler.trackRepeating) "looped" else "unlooped"),
            Pair("{title}", track.title),
            Pair("{author}", track.author)
        )

        return embed
    }

    private fun handleQueueRepeat(musicManager: GuildMusicManager, looper: User): MessageEmbed {
        val scheduler = musicManager.scheduler
        val queueHandler = scheduler.queueHandler
        val guild = musicManager.guild
        val player = musicManager.player
        val embed = if (queueHandler.queueRepeating) {
            queueHandler.queueRepeating = false
            queueHandler.clearSavedQueue()
            RobertifyEmbedUtils.embedMessage(guild, LoopMessages.QUEUE_LOOP_STOP).build()
        } else {
            if (player.playingTrack == null) {
                RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.NOTHING_PLAYING)
                    .build()
            } else {
                val thisTrack = player.playingTrack
                if (queueHandler.isEmpty)
                    RobertifyEmbedUtils.embedMessage(guild, LoopMessages.QUEUE_LOOP_NOTHING)
                        .build()
                else {
                    queueHandler.queueRepeating = true
                    scheduler.addToBeginningOfQueue(thisTrack)
                    queueHandler.setSavedQueue(queueHandler.contents)
                    queueHandler.remove(thisTrack)
                    RobertifyEmbedUtils.embedMessage(guild, LoopMessages.QUEUE_LOOP_START)
                        .build()
                }
            }
        }

        LogUtilsKt(guild).sendLog(
            LogType.TRACK_LOOP,
            LoopMessages.QUEUE_LOOP_LOG,
            Pair("{user}", looper.asMention),
            Pair("{status}", if (queueHandler.queueRepeating) "looped" else "unlooped")
        )
        return embed
    }

    override val help: String
        get() = "Set the song being currently played or the queue to constantly loop"
}